### PR TITLE
Add a readme for text and vision convert script modal type

### PR DESCRIPTION
**Pull request by karan6181**
## Description of changes:
- Add a readme for NLP and CV convert script modal type.
- Update the conversion scripts with method comments.

## Issue #, if available:
CO-1677

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [x] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/text/convert/README.md [hunk 1]
```diff
@@ -2,16 +2,71 @@
 
 To use Streaming Dataset, We first convert the dataset from its native format to MosaicML's streaming dataset format (a collection of binary `.mds` files). Once in `.mds` format, we can store the dataset in a central location (filesystem, S3, GCS, etc.) and stream the data to any compute cluster, with any number of devices, and any number of CPU workers, and it all just works.
 
-## Converting C4 to streaming dataset .mds format
+Follow the below steps to convert the Natural Language Processing dataset into a streaming MDS format. Also checkout the supported [MDSWriter()](https://streaming.docs.mosaicml.com/en/latest/api_reference/generated/streaming.MDSWriter.html) parameters for advanced usage.
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:5]: Here we refer to "MDS" as a name but above we refer to it as a ".mds" file extension.  I know it's both but it we should be consistent.  I recommend using "MDS" when discussing it as a format and keeping the ".mds" extension to code snippets and examples.

Comment (bandish-shah) [reply streaming/text/convert/README.md:5]:
```suggestion
Check out steps below for information on converting common NLP datasets to MDS format.  Please see [MDSWriter()](https://streaming.docs.mosaicml.com/en/latest/api_reference/generated/streaming.MDSWriter.html) parameters for details on advanced usage.
```

File: streaming/text/convert/README.md [hunk 2]
```diff
@@ -2,16 +2,71 @@
...
-# For 'train' split, it will take anywhere from 1-to-many hours depending on bandwidth, # CPUs, etc.
-# Once the script completes, you should see a dataset folder `./my-copy-c4/val` that is ~0.5GB and
-# `./my-copy-c4/train` this is ~800GB.
+## 1. [C4: Colossal, Cleaned, Common Crawl dataset](https://huggingface.co/datasets/c4)
 
-python c4.py --out_root ./my-copy-c4
-```
+**Instructions:**
+
+1. Run the [c4.py](c4.py) script as shown below with the minimalist arguments. The script downloads the raw format `train` and `val` split from HuggingFace hub and converts to StreamingDataset MDS format into their own split directory. For advanced users, you can look at the supported arguments for [c4.py](c4.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:12]:
```suggestion
1. Run the [c4.py](c4.py) script as shown below with the minimalist arguments. The script downloads the raw format with `train` and `val` splits from HuggingFace hub and converts to StreamingDataset MDS format into their own split directories. For more advanced use cases, please see the supported arguments for [c4.py](c4.py) and modify as necessary.
```

File: streaming/text/convert/README.md [hunk 3]
```diff
@@ -2,16 +2,71 @@
...
+    ```
+    python c4.py --out_root <local or remote directory path to save output MDS shard files>
+    ```
+
+## 2. [Wikipedia](https://huggingface.co/datasets/wikipedia)
+
+**Instructions:**
+
+1. Download English Wikipedia 2020-01-01 from [here](https://drive.google.com/drive/folders/1cywmDnAsrP5-2vsr8GDc6QUc7VWe-M3v).
+2. unzip the file `results_text.zip` as shown below.
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:23]:
```suggestion
2. Unzip the file `results_text.zip` as shown below.
```

File: streaming/text/convert/README.md [hunk 4]
```diff
@@ -2,16 +2,71 @@
...
+    <!--pytest.mark.skip-->
+    ```bash
+    unzip results_text.zip
+    ```
+
+    That will result in this directory structure:
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:29]:
```suggestion
    Listing the output should show the following directory structure:
```

File: streaming/text/convert/README.md [hunk 5]
```diff
@@ -2,16 +2,71 @@
...
+    ├── eval.txt
+    ├── part-00000-of-00500
+    ├── part-00001-of-00500
+    ├── part-00002-of-00500
+    ├── .....
+    ├── part-00498-of-00500
+    └── part-00499-of-00500
+    ```
+
+3. Run the [enwiki_text.py](enwiki_text.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset split into their own split directory. For advanced users, you can look at the supported arguments for [enwiki_text.py](enwiki_text.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:41]:
```suggestion
3. Run the [enwiki_text.py](enwiki_text.py) script with the minimalist arguments. The script converts the `train` and `val` dataset splits into their own split directories. For more advanced use cases, please see the supported arguments for [enwiki_text.py](enwiki_text.py) and modify as necessary.
```

File: streaming/text/convert/README.md [hunk 6]
```diff
@@ -2,16 +2,71 @@
...
+    ```
+    python enwiki_text.py --in_root <Above directory> --out_root <local or remote directory path to save output MDS shard files>
+    ```
+
+## 3. [Pile](https://pile.eleuther.ai/)
+
+**Instructions:**
+1. Download the Pile dataset from [here](https://the-eye.eu/public/AI/pile/).
+
+   That will result in this directory structure:
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:52]:
```suggestion
   Listing the output should show the following directory structure:
```

File: streaming/text/convert/README.md [hunk 7]
```diff
@@ -2,16 +2,71 @@
...
+    │   ├── 01.jsonl.zst
+    │   ├── 02.jsonl.zst
+    │   ├── 03.jsonl.zst
+    │   ├── .....
+    │   ├── 28.jsonl.zst
+    │   └── 29.jsonl.zst
+    └── val.jsonl.zst
+    ```
+
+2. Run the [pile.py](pile.py) script as shown below with the minimalist arguments. The script converts the `train`, `test`, and `val` dataset split into their own split directory. For advanced users, you can look at the supported arguments for [pile.py](pile.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:68]:
```suggestion
2. Run the [pile.py](pile.py) script with the minimalist arguments. The script converts the `train`, `test`, and `val` dataset splits into their own split directories. For more advanced use cases, yplease see the supported arguments for [pile.py](pile.py) and modify as necessary.
```

File: streaming/text/convert/README.md [hunk 8]
```diff
@@ -2,16 +2,71 @@
 
 To use Streaming Dataset, We first convert the dataset from its native format to MosaicML's streaming dataset format (a collection of binary `.mds` files). Once in `.mds` format, we can store the dataset in a central location (filesystem, S3, GCS, etc.) and stream the data to any compute cluster, with any number of devices, and any number of CPU workers, and it all just works.
```
Comment (bandish-shah) [inline streaming/text/convert/README.md:3]:
```suggestion
To use Streaming Dataset we must first convert the dataset from its native format to MosaicML's Streaming Dataset (MDS) format. Once in MDS format, we can access the from the local file system ( disk network attached storage, etc.) or object store (GCS, OCS, S3, etc.).  From object store, data can be streamed to train deep learning models and it all just works efficiently.
```

File: streaming/vision/convert/README.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
+# Dataset preparation
+
+To use Streaming Dataset, We first convert the dataset from its native format to MosaicML's streaming dataset format (a collection of binary `.mds` files). Once in `.mds` format, we can store the dataset in a central location (filesystem, S3, GCS, or OCI) and stream the data to any compute cluster, with any number of devices, and any number of CPU workers, and it all just works.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:3]:
```suggestion
To use Streaming Dataset we must first convert the dataset from its native format to MosaicML's Streaming Dataset (MDS) format. Once in MDS format, we can access the from the local file system ( disk network attached storage, etc.) or object store (GCS, OCS, S3, etc.).  From object store, data can be streamed to train deep learning models and it all just works efficiently.
```

File: streaming/vision/convert/README.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+Follow the below steps to convert the Computer Vision dataset into a streaming MDS format. Also checkout the supported [MDSWriter()](https://streaming.docs.mosaicml.com/en/latest/api_reference/generated/streaming.MDSWriter.html) parameters for advanced usage.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:5]:
```suggestion
Check out steps below for information on converting common Computer Vision datasets to MDS format. Please see [MDSWriter()](https://streaming.docs.mosaicml.com/en/latest/api_reference/generated/streaming.MDSWriter.html) parameters for details on advanced usage.
```

File: streaming/vision/convert/README.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+## 1. [ADE20K](https://groups.csail.mit.edu/vision/datasets/ADE20K/)
+
+**Instructions:**
+
+1. Download the ADE20K dataset from [here](https://groups.csail.mit.edu/vision/datasets/ADE20K/).
+2. That will result in this directory structure:
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:12]:
```suggestion
2. Listing the output should show the following directory structure:
```

File: streaming/vision/convert/README.md [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    ```bash
+    ├── annotations
+    │   ├── training
+    │   └── validation
+    └── images
+        ├── training
+        └── validation
+    ```
+
+3. Run the [ade20k.py](ade20k.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset split into their own directory. For advanced users, you can look at the supported arguments for [ade20k.py](ade20k.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:23]:
```suggestion
3. Run the [ade20k.py](ade20k.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset splits into their own directories. For advanced use cases, please see the supported arguments for [ade20k.py](ade20k.py) and modify according as necessary.
```

File: streaming/vision/convert/README.md [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+      <!--pytest.mark.skip-->
+      ```
+      python ade20k.py --in_root <Above directory> --out_root <local or remote directory path to save output MDS shard files>
+      ```
+
+## 2. [CIFAR10](https://www.cs.toronto.edu/~kriz/cifar.html)
+
+**Instructions:**
+
+1. Run the [cifar10.py](cifar10.py) script as shown below with the minimalist arguments which also downloads the raw CIFAR10 dataset if it doesn't exist. For advanced users, you can look at the supported arguments for [cifar10.py](cifar10.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:33]:
```suggestion
1. Run the [cifar10.py](cifar10.py) script as shown below with the minimalist arguments. The CIFAR10 dataset will be automatically downloaded if it doesn't exist locally. For advanced use cases, please see the supported arguments for [cifar10.py](cifar10.py) and modify as necessary.
```

File: streaming/vision/convert/README.md [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    unzip annotations_trainval2017.zip
+    unzip train2017.zip
+    unzip val2017.zip
+
+    rm annotations_trainval2017.zip
+    rm train2017.zip
+    rm val2017.zip
+    ```
+
+    That will result in this directory structure:
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:60]:
```suggestion
    Listing the output should show the following directory structure:
```

File: streaming/vision/convert/README.md [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    │   └── instances_val2017.json
+    ├── train2017
+    │   ├── 000000391895.jpg
+    |   |── ...
+    └── val2017
+    │   ├── 000000000139.jpg
+    |   |── ...
+    ```
+
+2. Run the [coco.py](coco.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset split into their own directory. For advanced users, you can look at the supported arguments for [coco.py](coco.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:74]:
```suggestion
2. Run the [coco.py](coco.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset splits into their own directories. For advanced use cases, please seet the supported arguments for [coco.py](coco.py) and modify as necessary.
```

File: streaming/vision/convert/README.md [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    <!--pytest.mark.skip-->
+    ```
+    python coco.py --in_root <Above directory> --out_root <local or remote directory path to save output MDS shard files>
+    ```
+
+## 4. [ImageNet](https://www.image-net.org/)
+
+**Instructions:**
+
+1. Download the ImageNet dataset from [here](https://image-net.org/download.php). More specifically, 2 files would be needed, `ILSVRC2012_img_train.tar` for training and `ILSVRC2012_img_val.tar` for validation. Then, untar both the files as shown below.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:84]:
```suggestion
1. Download the ImageNet dataset from [here](https://image-net.org/download.php). Two files are needed, `ILSVRC2012_img_train.tar` for training and `ILSVRC2012_img_val.tar` for validation. Next untar both the files as shown below.
```

File: streaming/vision/convert/README.md [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    tar -xvf ILSVRC2012_img_val.tar -C val/
+    rm ILSVRC2012_img_val.tar
+
+    mkdir train
+    mv ILSVRC2012_img_train.tar train/
+    tar -xvf ILSVRC2012_img_train.tar -C train/
+    rm ILSVRC2012_img_train.tar
+    ```
+
+    That will result in this directory structure:
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:98]:
```suggestion
    Listing the output should show the following directory structure:
```

File: streaming/vision/convert/README.md [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+      ├── ......
+    ├── val/
+      ├── n01440764
+      │   ├── ILSVRC2012_val_00000293.JPEG
+      │   ├── ILSVRC2012_val_00002138.JPEG
+      │   ├── ......
+      ├── ......
+    ```
+
+2. Run the [imagenet.py](imagenet.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset split into their own directory. For advanced users, you can look at the supported arguments for [imagenet.py](imagenet.py) and change according to your own needs.
```
Comment (bandish-shah) [inline streaming/vision/convert/README.md:115]:
```suggestion
2. Run the [imagenet.py](imagenet.py) script as shown below with the minimalist arguments. The script converts the `train` and `val` dataset splits into their own directories. For advanced uses cases, please see the supported arguments for [imagenet.py](imagenet.py) and modify as needed.
```

**Review by bandish-shah — CHANGES_REQUESTED** [review]
Made some minor grammatical and wording nits.  Will leave it up to @knighton to sign off on functionality.

**Comment by bandish-shah** [comment]
Would it also be possible to link to the README's on the docs site?  My impression is that these examples are nested in the `streaming` package and it's not obvious where to find them without digging.  We may want to eventually consider up-leveling the conversion scripts in the repo, it doesn't really make sense to have these scripts in the package unless users can invoke them through the `streaming` API for conversion.

File: streaming/text/convert/README.md (hunk 2 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md (hunk 3 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md (hunk 4 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md (hunk 5 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md (hunk 6 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md (hunk 7 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md (hunk 8 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 9 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 10 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 11 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 12 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 13 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 14 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 15 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 16 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 17 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/vision/convert/README.md (hunk 18 above)
Comment (karan6181) [reply streaming/vision/convert/README.md]: Addressed as part of https://github.com/mosaicml/streaming/pull/183/commits/4d242aaf5f26f73278a555a3021da6e328d0a28b.

File: streaming/text/convert/README.md [hunk 19]
```diff
@@ -1,17 +1,73 @@
 # Dataset preparation
 
-To use Streaming Dataset, We first convert the dataset from its native format to MosaicML's streaming dataset format (a collection of binary `.mds` files). Once in `.mds` format, we can store the dataset in a central location (filesystem, S3, GCS, etc.) and stream the data to any compute cluster, with any number of devices, and any number of CPU workers, and it all just works.
+To use Streaming Dataset we must first convert the dataset from its native format to MosaicML's Streaming Dataset format called Mosaic Data Shard (MDS). Once in MDS format, we can access the from the local file system ( disk network attached storage, etc.) or object store (GCS, OCS, S3, etc.).  From object store, data can be streamed to train deep learning models and it all just works efficiently.
```
Comment (knighton) [inline streaming/text/convert/README.md:3]: i suppose Mosaic Dataset Shard is clearer?

spurious space "( "

File: streaming/text/convert/README.md [hunk 20]
```diff
@@ -1,17 +1,73 @@
...
-# For 'train' split, it will take anywhere from 1-to-many hours depending on bandwidth, # CPUs, etc.
-# Once the script completes, you should see a dataset folder `./my-copy-c4/val` that is ~0.5GB and
-# `./my-copy-c4/train` this is ~800GB.
+## 1. [C4: Colossal, Cleaned, Common Crawl dataset](https://huggingface.co/datasets/c4)
 
-python c4.py --out_root ./my-copy-c4
-```
+**Instructions:**
+
+1. Run the [c4.py](c4.py) script as shown below with the minimalist arguments. The script downloads the raw format with `train` and `val` splits from HuggingFace hub and converts to StreamingDataset MDS format into their own split directories. For more advanced use cases, please see the supported arguments for [c4.py](c4.py) and modify as necessary.
```
Comment (knighton) [inline streaming/text/convert/README.md:12]: calling it minimalist is not very minimalist

**Review by knighton — APPROVED** [review]

File: streaming/text/convert/README.md (hunk 19 above)
Comment (karan6181) [reply streaming/text/convert/README.md]: Addressed in https://github.com/mosaicml/streaming/pull/183/commits/c022ee4c491d76088cbd44f4c6f39f865a46ad9c.

**Comment by karan6181** [comment]
> Would it also be possible to link to the README's on the docs site? My impression is that these examples are nested in the `streaming` package and it's not obvious where to find them without digging. We may want to eventually consider up-leveling the conversion scripts in the repo, it doesn't really make sense to have these scripts in the package unless users can invoke them through the `streaming` API for conversion.

Yes, valid point. I have added a short and concise doc `dataset_conversion_to_mds.md` that references the readme.

**Review by bandish-shah — APPROVED** [review]
👍

**Pull request merged**